x/stake/types/validator.go

```diff
@@ -234,145 +299,102 @@ func (v Validator) ABCIValidatorZero() abci.Validator {
+			pool = pool.looseTokensToBonded(v.Tokens)
 		}
-		pool, tokens = pool.removeSharesUnbonded(v.PoolShares.Amount)
-
 	case sdk.Unbonding:
-		if NewStatus == sdk.Unbonding {
+
+		switch NewStatus {
+		case sdk.Unbonding:
 			return v, pool
+		case sdk.Bonded:
+			pool = pool.looseTokensToBonded(v.Tokens)
 		}
-		pool, tokens = pool.removeSharesUnbonding(v.PoolShares.Amount)
-
 	case sdk.Bonded:
-		if NewStatus == sdk.Bonded {
-			// Return if nothing needs switching
+
+		switch NewStatus {
+		case sdk.Bonded:
 			return v, pool
+		default:
+			pool = pool.bondedTokensToLoose(v.Tokens)
 		}
-		pool, tokens = pool.removeSharesBonded(v.PoolShares.Amount)
-	}
-
-	switch NewStatus {
-	case sdk.Unbonded:
-		pool, v.PoolShares = pool.addTokensUnbonded(tokens)
-	case sdk.Unbonding:
-		pool, v.PoolShares = pool.addTokensUnbonding(tokens)
-	case sdk.Bonded:
-		pool, v.PoolShares = pool.addTokensBonded(tokens)
 	}
 
+	v.Status = NewStatus
 	return v, pool
 }
 
-// RemovePoolShares removes pool shares from a validator. It returns
-// corresponding tokens, which could be burned (e.g. when slashing a validator)
-// or redistributed elsewhere.
-func (v Validator) RemovePoolShares(pool Pool, poolShares sdk.Rat) (Validator, Pool, int64) {
-	var tokens int64
-
-	switch v.Status() {
-	case sdk.Unbonded:
-		pool, tokens = pool.removeSharesUnbonded(poolShares)
-	case sdk.Unbonding:
-		pool, tokens = pool.removeSharesUnbonding(poolShares)
-	case sdk.Bonded:
-		pool, tokens = pool.removeSharesBonded(poolShares)
+// removes tokens from a validator
+func (v Validator) RemoveTokens(pool Pool, tokens sdk.Rat) (Validator, Pool) {
+	if v.Status == sdk.Bonded {
+		pool = pool.bondedTokensToLoose(tokens)
 	}
 
-	v.PoolShares.Amount = v.PoolShares.Amount.Sub(poolShares)
-	return v, pool, tokens
-}
-
-// EquivalentBondedShares ...
-//
-// TODO: Remove should only be tokens get the power or potential power for a
-// validator if bonded, the power is the BondedShares if not bonded, the power
-// is the amount of bonded shares which the the validator would have it was
-// bonded.
-func (v Validator) EquivalentBondedShares(pool Pool) (eqBondedShares sdk.Rat) {
-	return v.PoolShares.ToBonded(pool).Amount
+	v.Tokens = v.Tokens.Sub(tokens)
+	return v, pool
 }
 
 //_________________________________________________________________________________________________________
 
 // AddTokensFromDel adds tokens to a validator
 func (v Validator) AddTokensFromDel(pool Pool, amount int64) (Validator, Pool, sdk.Rat) {
-	var (
-		poolShares             PoolShares
-		equivalentBondedShares sdk.Rat
-	)
 
 	// bondedShare/delegatedShare
-	exRate := v.DelegatorShareExRate(pool)
+	exRate := v.DelegatorShareExRate()
 
-	switch v.Status() {
-	case sdk.Unbonded:
-		pool, poolShares = pool.addTokensUnbonded(amount)
-	case sdk.Unbonding:
-		pool, poolShares = pool.addTokensUnbonding(amount)
-	case sdk.Bonded:
-		pool, poolShares = pool.addTokensBonded(amount)
+	if v.Status == sdk.Bonded {
```

Comment:
`amountRat` can be declared a line earlier to avoid the extra `sdk.Rat` allocation

Reply:
addressed